File: Python/merge_sort.py
import random
import time
import matplotlib.pyplot as plt


# Merge Sort function
def merge_sort(arr):
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    return merge(left, right)


# Function to merge two halves
def merge(left, right):
    result = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])
    return result


# Different sizes of input
sizes = [1000, 2000, 4000, 8000, 16000]
times = []

for n in sizes:
    arr = [random.randint(1, 100000) for _ in range(n)]

    start = time.time()
    merge_sort(arr)  # call sorting function
    end = time.time()

    taken = end - start
    times.append(taken)
    print(f"{n} elements sorted in {taken:.6f} seconds")

# Plot graph
plt.plot(sizes, times, marker="o")
plt.xlabel("Number of Elements (n)")
plt.ylabel("Time Taken (seconds)")
plt.title("Merge Sort Performance")
plt.show()

File: C++ DSA/merge_sort.cpp
#include <iostream>
using namespace std;

void merge(int a[], int l, int m, int r) {
    int i=l, j=m+1, k=0, temp[r-l+1];
    while(i<=m && j<=r) temp[k++] = (a[i]<a[j]) ? a[i++] : a[j++];
    while(i<=m) temp[k++] = a[i++];
    while(j<=r) temp[k++] = a[j++];
    for(i=0;i<k;i++) a[l+i]=temp[i];
}

void mergeSort(int a[], int l, int r) {
    if(l<r) {
        int m=(l+r)/2;
        mergeSort(a,l,m);
        mergeSort(a,m+1,r);
        merge(a,l,m,r);
    }
}

int main() {
    int a[]={5,2,8,4,1}, n=5;
    mergeSort(a,0,n-1);
    for(int x:a) cout<<x<<" ";
}
